### PR TITLE
JAVA-971: Make type codecs invariant.

### PR DESCRIPTION
**Pull request by adutra**
This commit makes TypeCodec.accepts(TypeToken) invariant with respect to the Java type passed as argument.
TypeCodec.accepts(Object) remains covariant with respect to the runtime type of the passed object.
This commit also updates the documentation on custom codecs.

### CODE REVIEW

File: driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java [hunk 1]
```diff
@@ -508,6 +508,14 @@ public DataType getCqlType() {
     /**
      * Return {@code true} if this codec is capable of serializing
      * the given {@code javaType}.
+     * <p>
+     * The default implementation is <em>invariant</em> with respect to the passed
+     * argument (through the usage of {@link TypeToken#equals(Object)}.
+     * This means that, by default, a codec will only ever accept the
+     * <em>exact</em> Java type that it has been created for.
```
Comment (olim7t) [inline driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java:516]: "by default" implies that you can override it. I can't think of any legitimate case where you would do something different in a subclass (without wreaking havoc), so maybe this should be more like "this is invariant and it's strongly recommended not to override it". Or maybe the method should even be final...

File: features/custom_codecs/README.md [hunk 2]
```diff
@@ -374,6 +374,212 @@ Foo<Bar> foo = row.get(0, new TypeToken<Foo<Bar>>(){})
 Foo<Bar> foo = row.get(0, Foo.class);
 ```
 
+### Support for subtype polymorphism
+
+Suppose the following class hierarchy:
+
+```java
+class A {}
+class B extends A {}
```
Comment (olim7t) [inline features/custom_codecs/README.md]: Would be nice to have more expressive class names here as well.

File: features/custom_codecs/README.md [hunk 3]
```diff
@@ -374,6 +374,212 @@ Foo<Bar> foo = row.get(0, new TypeToken<Foo<Bar>>(){})
...
+Another anti-pattern that is strongly discouraged is to register a codec for
+a Java type and another codec for one of its superclasses or superinterfaces.
+In this situation, the registry will choose the first registered one,
+*even if the other one is a better match*.
+
+For example, if there are registered codecs for both `A` and `B`
+(i.e. `ACodec extends TypeCodec<A>` and `BCodec extends TypeCodec<B>`),
+when looking for a codec accepting `A` instances, which codec the registry will choose
+will depend on their registration order. This can lead to unpredictable results
+and runtime errors, and thus should always be avoided.
```
Comment (olim7t) [inline features/custom_codecs/README.md]: Well that's only true if you search for a codec for a value, right? Now that the type-based methods are invariant, B.class will return the BCodec and A.class the ACodec, whatever the order.

File: features/custom_codecs/README.md [hunk 4]
```diff
@@ -374,6 +374,212 @@ Foo<Bar> foo = row.get(0, new TypeToken<Foo<Bar>>(){})
...
+
+#### Strategies for mapping polymorphic data models
```
Comment (olim7t) [inline features/custom_codecs/README.md]: Frankly I would leave out this section altogether. That's a lot to digest for someone reading the doc for the first time, I don't think we should actively encourage polymorphism and if users want to get into it the "one codec per class hierarchy" should be intuitive enough (and the other approach is not really polymorphism, it's just different codecs for different columns).

File: driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java (hunk 1 above)
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java:516]: Good call, I rephrased the javadocs as you suggested.

File: features/custom_codecs/README.md (hunk 2 above)
Comment (adutra) [inline features/custom_codecs/README.md]: Now using the ubiquitous Animal / Cat example.

File: features/custom_codecs/README.md (hunk 3 above)
Comment (adutra) [inline features/custom_codecs/README.md]: I removed these last 2 paragraphs altogether.

File: features/custom_codecs/README.md (hunk 4 above)
Comment (adutra) [inline features/custom_codecs/README.md]: Section removed.

File: driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java [hunk 5]
```diff
@@ -548,13 +557,17 @@ public boolean accepts(DataType cqlType) {
...
+     * This means that, by default, a codec will accept
+     * <em>any subtype</em> of the Java type that it has been created for.</li>
      * <li>The base implementation provided here can only handle non-parameterized types;
      * codecs handling parameterized types, such as collection types, must override
      * this method and perform some sort of "manual"
      * inspection of the actual type parameters.</li>
-     * <li>Similarly, codecs that only accept a partial subset of all possible values,
-     * such as {@link TimeUUIDCodec} or {@link AsciiCodec}, must override
-     * this method and manually inspect the object to check if it
+     * <li>Similarly, codecs that only accept a partial subset of all possible values
```
Comment (adutra) [inline driver-core/src/main/java/com/datastax/driver/core/TypeCodec.java:569]: This is actually an oversight from JAVA-965, we shouldn't be mentioning AsciiCodec and TimeUUIDCodec as examples of codecs accepting only certain values, as it is not the case anymore.

File: features/custom_codecs/README.md [hunk 6]
```diff
@@ -374,6 +374,45 @@ Foo<Bar> foo = row.get(0, new TypeToken<Foo<Bar>>(){})
 Foo<Bar> foo = row.get(0, Foo.class);
 ```
 
+### Support for subtype polymorphism
```
Comment (tolbertam) [inline features/custom_codecs/README.md:377]: I really like the explanation here.  It is very concise and helps me understand how this works.

File: features/object_mapper/creating/README.md [hunk 7]
```diff
@@ -43,9 +43,9 @@ The class must provide a default constructor, and all fields (except the
 ones annotated `@Transient`) must have corresponding setters and
 getters.
 
-[table]:http://docs.datastax.com/en/drivers/java/2.2/com/datastax/driver/mapping/annotations/Table.html
-[case-sensitive]:http://docs.datastax.com/en/cql/3.2/cql/cql_reference/ucase-lcase_r.html
-[consistency level]:http://docs.datastax.com/en/drivers/java/2.2/com/datastax/driver/core/ConsistencyLevel.html
+[table]:http://docs.datastax.com/en/drivers/java/3.0/com/datastax/driver/mapping/annotations/Table.html
+[case-sensitive]:http://docs.datastax.com/en/cql/3.1/cql/cql_reference/ucase-lcase_r.html
```
Comment (tolbertam) [inline features/object_mapper/creating/README.md]: You'll probably encounter this in the rebase, but this has been updated to 3.3 from the upstream branches (2.0/2.1) and that seems to work good too.

**Comment by tolbertam** [comment]
Looks good to me :+1:

**Pull request merged**